File: PHP/src/Parser/Parser.php
<?php

declare(strict_types=1);

namespace Shin1x1\ToyJsonParser\Parser;

use Shin1x1\ToyJsonParser\Lexer\Lexer;
use Shin1x1\ToyJsonParser\Lexer\Token\EofToken;
use Shin1x1\ToyJsonParser\Parser\Exception\ParserException;

final class Parser
{
    public function __construct(private readonly Lexer $lexer)
    {
    }

    public function parse(): array|string|int|float|bool|null
    {
        $ret = ValueParser::parse($this->lexer, $this->lexer->getNextToken());

        if ($this->lexer->getNextToken() instanceof EofToken) {
            return $ret;
        }

        throw new ParserException(message: 'Unparsed tokens detected');
    }
}

File: Kotlin/src/main/kotlin/parser/Parser.kt
package parser

import lexer.Lexer
import stream.EotException

class Parser(
    private val lexer: Lexer,
) {
    fun parse(): Result<JsonValue> {
        if (lexer.isEot()) {
            return Result.success(JsonValue.Null)
        }

        val json =
            lexer.getNextToken().fold(
                onSuccess = { token ->
                    ValueParser.parse(lexer, token).onFailure { return Result.failure(it) }
                },
                onFailure = { return Result.failure(it) },
            )

        return lexer.getNextToken().fold(
            onSuccess = { Result.failure(Exception("Some tokens are remained: $it")) },
            onFailure = { if (it is EotException) json else Result.failure(it) },
        )
    }
}
